src/services/OswServiceTaskNotifier.cpp

```diff
@@ -0,0 +1,127 @@
+#include "./services/OswServiceTaskNotifier.h"
+
+unsigned Notification::count = 0;
+
+NotificationData OswServiceTaskNotifier::createNotification(std::chrono::time_point<std::chrono::system_clock, std::chrono::seconds> timeToFire, std::string publisher,
+                                                            std::string message, std::array<bool, 7> daysOfWeek)
+{
+    const std::lock_guard<std::mutex> lock{mutlimapMutex};
+    auto notification = Notification{publisher, message, daysOfWeek};
+    auto pair = std::make_pair(timeToFire, notification);
+    scheduler.insert(pair);
+    return pair;
+}
+
+std::chrono::time_point<std::chrono::system_clock, std::chrono::seconds> getTimeToFire(int hours, int minutes)
+{
+    auto utcTime = std::chrono::system_clock::from_time_t(OswHal::getInstance()->getUTCTime());
+    auto currentTime = utcTime + std::chrono::seconds{static_cast<int>((OswConfigAllKeys::timeZone.get() + OswConfigAllKeys::daylightOffset.get()) * 3600)};
+    auto currentDate = date::sys_days(floor<date::days>(currentTime));
+    auto scheduledTime = std::chrono::hours{hours} + std::chrono::minutes{minutes} + std::chrono::seconds{0};
+    auto timeToFire = currentDate + scheduledTime;
+    if (currentTime >= timeToFire)
+    {
+        timeToFire += date::days{1};
+    }
+    return timeToFire;
+}
+
+// At least one day of week must be selected!
+std::chrono::time_point<std::chrono::system_clock, std::chrono::seconds> getTimeToFire(int hours, int minutes, std::array<bool, 7> daysOfWeek)
+{
+    auto timeToFire = getTimeToFire(hours, minutes);
+    auto scheduledDays = date::sys_days{floor<date::days>(timeToFire)};
+    auto dayOfWeek = date::weekday{scheduledDays};
+    auto i = dayOfWeek.c_encoding();
+    while (!daysOfWeek[i])
+    {
+        timeToFire += date::days{1};
+        i = (i + 1) % daysOfWeek.size();
+    }
+    return timeToFire;
+}
+
+NotificationData OswServiceTaskNotifier::createNotification(int hours, int minutes, std::string publisher,
+                                                            std::string message, std::array<bool, 7> daysOfWeek)
+{
+    const std::lock_guard<std::mutex> lock{mutlimapMutex};
+    std::chrono::time_point<std::chrono::system_clock, std::chrono::seconds> timeToFire{};
+    auto notification = Notification{publisher, message, daysOfWeek};
+    if (std::any_of(daysOfWeek.begin(), daysOfWeek.end(), [](auto x)
+                    { return x; }))
+    {
+        timeToFire = getTimeToFire(hours, minutes, daysOfWeek);
+    }
+    else
+    {
+        timeToFire = getTimeToFire(hours, minutes);
+    }
+    auto pair = std::make_pair(timeToFire, notification);
+    scheduler.insert(pair);
+    return pair;
+}
+
+std::vector<NotificationData> OswServiceTaskNotifier::readNotifications(const std::string publisher)
+{
+    const std::lock_guard<std::mutex> lock{mutlimapMutex};
+    std::vector<NotificationData> result{};
```

Comment:
Consider changing this to a list, as you do not need to waste reallocation cycles here :)

Reply:
I'd still use a vector as the default container. We only need to increase the vector's capacity once, as we know it won't ever hold more notifications than the multimap has.

Reply:
Oh, fixed in 99f6bf4a912f24118bee9159f61df921c90a8094 -> ignore me